### DISCUSSION TITLE
"Installation" of compiled mitsuba (windows)

### DISCUSSION THREAD

**Discussion by ssh4net** [mitsuba-renderer/mitsuba3#161] [Q&A]
Sorry for pretty dumb request, but is it possible add "how to" install compiled Mitsuba to make it work as installed with pip?
For this moment it required always run setpath.bat before any works in Jupyter notebook.

**Comment by merlinND** [comment]
Hello @ssh4net,

I haven't personally tested this, but in principle it should be as easy as:

```
python3 ./setup.py install
```

Another way would be to keep everything in place inside of your `mitsuba3/build` folder, and then modify your system-wide environment variables to point to that directory (essentially doing the same as `setpath.bat`, but persistent and global).

**Comment by anderslanglands** [comment]
On this topic, the cmake install *almost* works. The following changes need to be made to make it work ootb. Assuming the following sequence of commands:

```bash
git clone <mitsuba repo> mitsuba3
cd mitsuba3
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX="../install"
cmake --build . --config Release --target install
```

This installs most of the pieces, but is missing drjit. To make it work I had to then:
```bash
cp -r ..\ext\drjit\drjit ../install
cp .\Release\python\drjit\drjit_ext.cp39-win_amd64.pyd ../install/drjit
```

And finally I needed to modify drjit's `__init__.py` to add the installed path to the DLLS:
```python
if os.name == 'nt':
    os.add_dll_directory(os.path.join(os.getenv("MITSUBA_DIR"), "bin"))
```

This still isn't quite enough to build against the installed mitsuba distribution: the drjit headers are also missing but I haven't needed or had time to look into that yet.

**Comment by njroussel** [comment]
For completeness, let me add the following:

As a general word of precaution, uninstall any `drjit` or `mitsuba` package you have downloaded through PyPi before using and installing your own builds. 

If you're simply looking to apply some modifications to mitsuba and then have it accessible through Python without any `PATH` or `PYTONPATH` modifications. You can simple do this:
`pip install /some/path/to/mitsuba`

This will also download any requirement from PyPi, including DrJit. If you want to roll your own version of Drjit, you must first (before installing mitsuba) run:
`pip install /some/path/to/drjit`

Both of these commands will place the packages in the appropriate `site-packages` folder of your Python environment. These commands should also, in most cases, create the proper folder hierarchy for `mitsuba` to find `drjit`'s libraries.
If you're installing your builds directly through `cmake`, you will need to carefully select your install paths to handle this. @anderslanglands's answer has an example of a small fix needed with his specific setup.

**Reply by ommaury** [reply]
Hi @njroussel, I'm not quite sure how to read your last sentence about @anderslanglands comment: I'm finding I have to do the same (i.e. manual copy after cmake install) or use the compilation output directly as my working dir (which contains drjit). Is that to be expected?

This is really not a high priority btw, I'm just curious if there's another way to configure cmake that would make the install work. Thanks!

**Reply by njroussel** [reply]
@ommaury I wanted to point out that there are a few extra (undocumented) steps, like the copying. 

Honestly, I don't see any reason to install `mitsuba/drjit` through CMake. If you're just running the code, using `pip` is much more reliable.

If you want to build againt `mitsuba`, I'm pretty sure you can also use the `pip` install and change your build system to pickup the `pip` install directories. The advantage is that the `pip` installations have all the headers and have the correct folder structure.
This was our initial intention: packaging `mitsuba` as a Python package which people would also be able to build against. Unfortunately, we never really tested it thoroughly. I'd be very happy if you could report back any issues or unexpected extra steps you had to take to make this work if you decide to go down this route :smile:

**Reply by ommaury** [reply]
Thanks @njroussel, I understand.
